### PR TITLE
Adds TOTP generator to the hash feature

### PR DESCRIPTION
**Pull request by adamsong**
Allows for getting TOTP key from key, first time coding in rust so the code is probably bad, accepting any suggestions for how to make it less bad.

### CODE REVIEW

**Comment by willox** [comment]
You might want a way to generate a couple of tokens around the current one to add a little leniency for clients with sightly out of sync system times.

File: src/hash.rs
Comment (willox) [inline src/hash.rs:85]: We already use the [hex](https://crates.io/crates/hex) crate in this module which could decode this input directly into to your array with `hex::decode_to_slice`.

File: src/hash.rs
Comment (willox) [inline src/hash.rs:93]: Since Rust 1.55.0 (which our CI is using) you can replace the loop and declarations here with:
```rust
let mut ipad = seed.map(|x| x ^ 0x36);
let mut opad = seed.map(|x| x ^ 0x5C);
```

I'm also not sure these need to be mutable.

File: src/hash.rs
Comment (willox) [inline src/hash.rs:120]: Same story as the rest really. I think you can just get rid of this entire array (and through doing so get rid of the loop.)

**Review by willox — CHANGES_REQUESTED** [review]

File: src/hash.rs
Comment (willox) [inline src/hash.rs:106]: You _could_ get rid of this loop by using `slice::clone_from_slice` like so:

```rust
    let mut ipad_time: [u8; 72] = [0; 72];
    ipad_time[..64].clone_from_slice(&ipad[..]);
    ipad_time[64..].clone_from_slice(&time_bytes[..]);
```

**But**, I think this ipad_time array (and hence the copy operations) shouldn't be necessary at all. You can pass the data into the Sha1 hasher using multiple calls to `Sha1::update`.

Here's an example of that in the Rust playground. It's using Sha2 because that's all that is available there: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ab98d46d41b57af58e2d6c740936149

**Review by willox — CHANGES_REQUESTED** [review]

**Comment by adamsong** [comment]
The code definitely looks cleaner now, thanks for the help

File: src/hash.rs
Comment (willox) [inline src/hash.rs:94]: Can't this decode directly into `seed`? I'm not seeing the reason to have both seed and seed_bytes.

File: src/hash.rs
Comment (willox) [inline src/hash.rs:82]: Perhaps should output valid JSON instead. I think we have some other functions that do this?

File: src/hash.rs
Comment (willox) [inline src/hash.rs:85]: Probably a debug message that needs removing?

**Review by willox — CHANGES_REQUESTED** [review]
Yep the code is no longer insanely intimidating at first glance. This is a great improvement.

File: src/hash.rs
Comment (adamsong) [reply src/hash.rs:94]: Seed is too long, decode_to_slice panics due to incorrect string length

Comment (willox) [reply src/hash.rs:94]: Could you pass a sub-slice of seed that is the correct length?

Comment (adamsong) [reply src/hash.rs]: That worked

**Comment by adamsong** [comment]
I left a test in there, I can remove it before this gets merged if you want me to though

**Comment by willox** [comment]
Can you add some sort of error handling? rust-g isn't great at this in general, but crashing on incorrect seed length seems a little too volatile

**Comment by adamsong** [comment]
I will see what I can do, is there an easy way to return an error to byond?

**Comment by willox** [comment]
Nope. All you've got to work with is strings.

You could use JSON again or just settle with something like returning an empty string on error.

**Comment by adamsong** [comment]
Would replacing most of the `unwrap` calls with `?` work?

**Comment by willox** [comment]
Yes. You might have to fiddle with the return value to get that to work though.

Returning None from inside your `byond_fn!` usage will result in an empty string being returned to BYOND. The `.ok()` you've got in there will turn an error into None.

**Comment by adamsong** [comment]
Some of these errors don't like being turned into `Error`, and I cannot set the error type on the result because `Result` has been redefined in `error.rs` to not take a second type parameter. I have a workaround that hopefully works, and some of the panics I am going to leave (If your system clock is past the year 200 billion, that is on you)

File: src/hash.rs
Comment (willox) [inline src/hash.rs:94]: You definitely want code like this to be closer to:

```rust
let result = totp_generate(hex_seed, i.try_into().unwrap(), time_override)?
results.push(result);
```

Which I assume is what you were actually going for before being blocked by the functions' Result type.

**Review by willox — CHANGES_REQUESTED** [review]
I'm not a fan of this approach.

I've had a look at what's going on in error.rs and it shouldn't be too hard to leverage. First of all just add some error messages into its Error enum here: https://github.com/tgstation/rust-g/blob/539c948a5132c3617045026f2fedfaeb23d5c99a/src/error.rs#L20-L58

Let's add:
```rust
    #[cfg(feature = "hash")]
    #[error(transparent)]
    InvalidSeed,
```

You can do the following to map errors from third-party APIs to your own:
```rust
fn totp_generate(hex_seed: &str, offset: i64, time_override: Option<i64>) -> Result<String> {
    let decode_result = hex::decode_to_slice(hex_seed, &mut seed[..10] as &mut [u8])
        .map_err(|_| Error::InvalidSeed)?;

    ...
}
```

There are also ways to automatically translate between errors of one type to another but I like this approach.

File: src/hash.rs
Comment (adamsong) [reply src/hash.rs:94]: The problem is I couldn't return a `FromHexError` and when I tried to instantiate `Error` I got an error I didn't understand, so I'm just returning an Ok, with the error message as the text.

Comment (adamsong) [reply src/hash.rs:94]: If I instead try to return `Err(String)` I get 'expected enum `error::Error`' and if I try `Err(Error::new)` I get 'variant or associated item cannot be called on `error::Error` due to unsatisfied trait bounds'

Comment (willox) [reply src/hash.rs:94]: With `Error::new` I assume you were trying to create an instance of the Error enum in our error.rs module? It's an enum so to create an instance you can pick one of the variants and use it directly.

For example our first variant is Null: https://github.com/tgstation/rust-g/blob/539c948a5132c3617045026f2fedfaeb23d5c99a/src/error.rs#L19-L22

To use that in some code you could do:
```rust
fn my_func() -> Result<u32> {
    return Err(Error::Null)
}
```

Comment (adamsong) [reply src/hash.rs:94]: Can I add my own error to the Enum? None of them fit very well

Comment (adamsong) [reply src/hash.rs]: I now see that is what you suggested in the original comment

**Review by willox — APPROVED** [review]

**Comment by willox** [comment]
I probably won't merge this before tomorrow.

**Pull request merged**
